src/proxy/processors/push-action/pullRemote.js

```diff
@@ -20,22 +20,43 @@ const exec = async (req, action) => {
     }
 
     const cmd = `git clone ${action.url} --bare`;
-    step.log(`Exectuting ${cmd}`);
 
-    // eslint-disable-next-line max-len
-    const response = execSync(`git clone ${action.url} --bare`, {
+    // Retrieve authorization headers
+    const authorizationHeader = req.headers?.authorization;
+
+    // Validate the authorization headers
+    const authorizationValid =
+      authorizationHeader &&
+      typeof authorizationHeader === 'string' &&
+      authorizationHeader.includes('Basic ');
+
+    // Construct clone URL depending on presence of authorization headers
+    const cloneUrl = authorizationValid
+      ? `https://${Buffer.from(authorizationHeader.split(' ')[1], 'base64')}@${action.url.replace(
+          /https*:\/\//,
+          '',
+        )}`
+      : action.url;
+
+    step.log(`Exectuting ${cmd}${authorizationValid ? ' with credentials' : ''}`);
+
+    const response = spawnSync('git', ['clone', cloneUrl, '--bare', '--progress'], {
```

Comment:
I was recently bit by an issue using `spawnSync` on another Express application. If this process doesn't return in a timely manner, it can block the whole event loop including Express's HTTP routing/handlers. For a large repo, it's fair to assume this could take dozens of seconds or minutes to complete. I was also wrapping a call to `git` CLI and used the below pattern instead. Not sure if this will work for our usage of it here since we still have to "wait" for the clone to finish as part of the pull.

```javascript
const promisify = require('util').promisify;
const execFile = require('child_process').execFile;
const execFileAsync = promisify(execFile);

const exec = async (req, action) => {
  ...
  const { stdout } = await execFileAsync('git', ['clone', cloneUrl, '--bare', '--progress'], {
    cwd: action.proxyGitPath,
  });
  console.log(stdout);
}
```

Reply:
I am not sure. If the function is asynchronous, we might be implementing a race condition where it is unclear whether the function call will end prior to the completion of the pull? I'd suggest that we leave it for now, but definitely open an issue to quickly swap out these `git` calls with a tried and tested library that serves as a wrapper around the `git` protocol.

Thoughts?

Reply:
You should be able to validate this behaviour with trying to pull down a large repository. We saw this on [`backstage/backstage`](https://github.com/backstage/backstage) which is a couple GiB's. I suspect the current implementation will cause the server to stop responding to other requests such as health checks or API requests as it blocks the whole event loop waiting for `spawnSync` to return.

The [`spawn`](https://nodejs.org/api/child_process.html#child_processspawncommand-args-options) function can also be used but it relies on callbacks and is a bit more tricky to implement correctly since the `<cmd>.stdout.on(...)` callbacks happen asynchronously 😫

Reply:
@coopernetes - note that we should definitely open an issue for this and address post merge 👍 Are you able to create an issue and reference?